### ISSUE TITLE
"Watch for Changes" does not work.

### ISSUE THREAD

**Issue by hardhub** [syncthing/syncthing#7944]
Nothing simply happens when I delete files on remote machine.

Rescan works and help.

Can it be incompatibility with DiscCryptor?

**Comment by bt90** [comment]
Diskcryptor uses FUSE under the hood? Quite likely that it doesn't implement the necessary functionality (yet):

https://github.com/libfuse/libfuse/wiki/Fsnotify-and-FUSE

If that's the case, you should open an issue on the diskcryptor tracker.

**Comment by hardhub** [comment]
> 
> 
> Diskcryptor uses FUSE under the hood? Quite likely that it doesn't implement the necessary functionality (yet):
> 
> https://github.com/libfuse/libfuse/wiki/Fsnotify-and-FUSE
> 
> If that's the case, you should open an issue on the diskcryptor tracker.

DiskCryptor is Windows only software. And it works well and transparent for the system, even shrinking works well.

I forgot to tell boths sides are Windows 7 x64 and NTFS.

I think I need to test on partition without FDE but for the moment I do not have any one.

So I think it can be not related... most likely. Which win 32 API used to track FS changes in Syncthing on Windows?

Probably I can run some small code to check if tracking happens. Or some logging options which can help?

**Comment by bt90** [comment]
Sorry, my bad. Confused it with [Cryptomator](https://cryptomator.org).

**Comment by tomasz1986** [comment]
> I forgot to tell boths sides are Windows 7 x64 and NTFS.

I can only add that this likely has nothing to do with Windows 7. I've just done a test under the OS, and file watching seems to work correctly in it.

**Comment by AudriusButkevicius** [comment]
This looks more like a support request than a clear bug or feature request. Please use the forum - https://forum.syncthing.net/ - for support, first by searching for any similar issues and then by posting a new one when required. The forum has a lot more eyes and friendly people on it than the issue tracker.

Thanks!

**Comment by hardhub** [comment]
> 
> 
> This looks more like a support request than a clear bug or feature request. Please use the forum - https://forum.syncthing.net/ - for support, first by searching for any similar issues and then by posting a new one when required. The forum has a lot more eyes and friendly people on it than the issue tracker.
> 
> Thanks!

@AudriusButkevicius I already wrote on forum and it is obviously bug.

**Comment by AudriusButkevicius** [comment]
It's a thing that is provided by the filesystem. 
I don't think its a bug if the underlying virtual filesystem doesn't support it.

**Comment by hardhub** [comment]
> It's a thing that is provided by the filesystem.
> I don't think its a bug if the underlying virtual filesystem doesn't support it.

Which virtual FS? NTFS?

Can you please read more carefully what I wrote because current level of attention makes me nervous.

It is bug, and **I guess I have found the reason** already.  
Synchronization is not performed because most likely the code tracks wrong path.  
If we need to synchronize root of disk, for example `D:\`, I assume Syncthing tries to track path `D:` not `D:\` because it trims trailing slash always (even if it is not needed).   

If you use FindFirstChangeNotification or something like that then you need to meet naming rules (https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file). So I guess you simply do not receive notification due to wrong name.

![image](https://user-images.githubusercontent.com/4306334/133078751-3bca7bca-2c20-4a08-9899-de90299e73ff.png)

I tried to synchronize subdirectories and it works fine. So I almost sure it is trailing slash bug.  

P.S. I hope security-related code is written better ))

**Comment by hardhub** [comment]
And in general such decision just to close issue only because you are not able to understand something is bad decision.  
I am not familiar with Syncthing code base, I cannot fix it and send merge request, or easy find the reason without developer help.  
I asked about debug level logs and the code which reads file system change notifications - no reaction.  

D1: Maybe help him in analyzing details of an issue?
D2: No, let's close it )))
D1: But our software at risk to be a piece of shit finally.
D2 No, no, we are the best.

**Comment by AudriusButkevicius** [comment]
In general, just saying "it doesn't work" doesn't help, as it's not clear what the cause is.

Until we figure out what needs fixing, the discussions are best had on the forum, as for all that I know, you might disappear and never reply on this issue ever again, leaving a dangling issue forever.

**Comment by hardhub** [comment]
> In general, just saying "it doesn't work" doesn't help

Yes, I know, but it is difficult to start from something if the single symptom is just "it does not work".

> Until we figure out what needs fixing, the discussions are best had on the forum

It is first what I did, and I replied the following:

> I already wrote on forum and it is obviously bug.

>  leaving a dangling issue forever.

I agree, it makes me annoying as well especially if it is big project. But I replied to bt90 immediately and asked ;
> Which win 32 API used to track FS changes in Syncthing on Windows?
> Probably I can run some small code to check if tracking happens.
> Or some logging options which can help?

It does not look like I want to dissappear ))

**Comment by AudriusButkevicius** [comment]
we use a fork of https://github.com/rjeczalik/notify which I think uses ReadDirectoryChangesW

**Comment by hardhub** [comment]
> we use a fork of https://github.com/rjeczalik/notify which I think uses ReadDirectoryChangesW

I am sorry I am not GO developer ))  
So if somebody can look into code and check what the name is passed to _notify_ that will help a lot.  

https://github.com/syncthing/syncthing/blob/d47745a86b00fe1122d8ebd5e17601aa1b7552b2/lib/fs/basicfs_watch.go#L46

I expect it is now without trailing slash (because it is so on UI) but Naming Conventions in Windows say:  
> Use a **backslash as required as part of volume names**, for example, the "C:\" in "C:\path\file" or the "\\server\share" in "\\server\share\path\file" for Universal Naming Convention (UNC) names. For more information about UNC names, see the Maximum Path Length Limitation section.

So

**Comment by tomasz1986** [comment]
I think it's a good idea to add https://forum.syncthing.net/t/no-istant-sync/17497 in here. Drive letters alone aren't the culprit. File watching doesn't seem to work at all in drive root paths, even if we use mount points or junctions/symlinks instead of the letters.

**Comment by SubhashBose** [comment]
I have been watching this issue for over a year and was hoping this will be fixed. But I don't see any discussion or progress on this. is this abandoned?

This is a very fundamental functionality of Syncthing which is broken for Windows, and so I expected this to be fixed soon. 

Why this is still pending? I am not a GO developer, but I believe this is not an issue with Windows Filesystem API, rather it is an issue with Syncthing only. Moreover, Syncthing-Trayzor seems to do the file-system watching correctly for drive root directories. 

So, I just wanted to bump this to see where this is heading (if at all).

**Comment by AudriusButkevicius** [comment]
Bumping doesn't do much. Its still an issue and its still open. Its not fixed because none of the people that are able to fix it feel about this strongly.

**Comment by rasa** [comment]
For the folder directory specification, are you using `D:\`, or simply `D:`? The latter isn't the root folder, it's the "current" directory on drive D:.

**Comment by SubhashBose** [comment]
The trailing slash is removed by syncthing. I have tried both by D:\ to just D: . Even if I add D:\ , syncthing removed the slash and makes it D:

**Comment by AudriusButkevicius** [comment]
Agreed, I think thats the issue. Someone should test it and propose a PR fixing it.

**Comment by rasa** [comment]
I couldn't find in the code where to make this fix. If someone can point the way, I can provide a PR.

**Comment by imsodin** [comment]
Don't have time for an in-depth search, and couldn't find it either. I'd probably start by adding a testcase to `TestCanonicalize` in `./lib/fs` to see if it strips the trailing slash. And/or if you have a windows system, enable `fs` debug logging and watch e.g. `C:\` to see if our hypothesis even holds (i.e. we do actually tell the notify library to watch `C:`) - there's still a small chance that `notify` is the culprit here.

**Comment by P110** [comment]
Just checking in, is this still open with no work-around other than "move it to a subfolder" ?

**Comment by P110** [comment]
Has anyone actually got a subfolder working? I can't seem to get windows to watch any syncthing folder that I use?

**Comment by P110** [comment]
Has anyone actually got a subfolder working? I can't seem to get windows to watch any syncthing folder that I use?

**Comment by VioletGiraffe** [comment]
I have come across this issue looking for why `FindFirstChangeNotification` doesn't detect changes in a drive's root. I can confirm that `FindFirstChangeNotification` requires paths **without** a trailing slash, unless it's the root, in which case it only works with a trailing slash. This is true both for regular paths and UNC paths.
`ReadDirectoryChangesW` likely behaves in the same way (otherwise you wouldn't be having issue, and they probably both use the same kernel facility under the hood).

**Comment by crackwitz** [comment]
I just encountered this behavior myself. I'm just here to beg for someone to please address this in some way. I'd rather sync the entire drive but my situation allows me to create shares for the handful of individual directories. I'd DIY but I'm entirely unfamiliar with Go and this code base.

**Comment by timkgh** [comment]
If indeed it is just an issue of ending the root path with `/` or `\`, it should not be a difficult fix for someone familiar with the code.

**Comment by crackwitz** [comment]
I just tried `ReadDirectoryChangesW` from Python's FFI, ctypes. Both `"X:"` and `"X:\\"` worked fine.

I have only tried `FindFirstChangeNotification` using the pywin32 package, and there it also works for both styles of path.

If you are using a library to handle all this, it might be an issue in that library.

**Comment by RapidTransit** [comment]
I'm getting bit by this too on Windows, I also noticed it doesn't sync open files, does Syncthing not use VSS?

**Comment by tomasz1986** [comment]
> I also noticed it doesn't sync open files, does Syncthing not use VSS?

It does not, so yes, you cannot sync locked files in Windows. However, this is completely unrelated to the issue at hand, which is about the file watcher.

**Comment by RapidTransit** [comment]
> > I also noticed it doesn't sync open files, does Syncthing not use VSS?
> 
> It does not, so yes, you cannot sync locked files in Windows. However, this is completely unrelated to the issue at hand, which is about the file watcher.

I was under the impression VSS included an API people hooked into for file watching, just doing a little research that is not the case.

**Comment by rasa** [comment]
My initial research has discovered this is either an [notify](https://github.com/syncthing/notify) issue, or a Windows issue. 

To start my debugging I removed https://github.com/syncthing/syncthing/issues/5578's hack:
https://github.com/syncthing/syncthing/blob/110e1ae6f9e773ba58e02c3de2e0b59ce81e6ae3/lib/fs/basicfs_windows.go#L360-L362, which isn't required anymore. Setting a folder's root to `Z:` still fails.

To be clear, calling
https://github.com/syncthing/syncthing/blob/d47745a86b00fe1122d8ebd5e17601aa1b7552b2/lib/fs/basicfs_watch.go#L46
with:
`Z:` (passed as `\\?\Z:\...`): fails
`Z:\` (passed as `\\?\Z:\...`): fails
`Z:\sub` (passed as `\\?\Z:\sub\...`): works

I did find a workaround: If I use the Disk Management app, and mount drive `Z:` as a path under another drive, say `C:\Z`, then notify works fine.

I've verified that notify's [`cleanpath()`](https://github.com/syncthing/notify/blob/c6b7342338d286bbc1c7959b03ecdec301ea90d0/util.go#L50)
is adding a trailing `\` to root folders, so I need to drill down deeper, but my time is limited for the next few days, so please be patient.

**Comment by rasa** [comment]
It appears this is a notify bug that was [identified](https://github.com/rjeczalik/notify/issues/148) in 2018, and a potential fix was [provided](https://github.com/lzp9421/notify/commit/f079f99b247d8a356453420e59b4715cb070377a) in 2022. Seems like a good place to start.

**Comment by rasa** [comment]
So https://github.com/lzp9421/notify/commit/f079f99b247d8a356453420e59b4715cb070377a didn't fix this issue. I hacked on the code a bit, and realized it's not going to be an easy fix. So I have built a framework of [GitHub actions](https://github.com/rasa/notify/actions) to build, test, and lint the code. 

The tests are pretty [reliable](https://github.com/rasa/notify/actions/workflows/test.yaml) on Windows, but on Linux they [fail](https://github.com/rasa/notify/actions/workflows/test.yaml) often. And on macOS, the [fail](https://github.com/rasa/notify/actions/workflows/test-macos.yaml) almost all the time, using [kqueue](https://github.com/rasa/notify/actions/workflows/test-macos-kqueue.yaml) or not. All three BSD variants [fail](https://github.com/rasa/notify/actions/workflows/test-bsd.yaml) all the time, too.

And using this [test-windows-root](https://github.com/rasa/notify/actions/workflows/test-windows-root.yaml) workflow, we have a good framework to fix this issue.

**Comment by MU7L3Y** [comment]
@rasa Is this still being actively worked on? I've just been bit by this as well.

**Comment by rasa** [comment]
@MU7L3Y I worked on it a bit [recently](https://github.com/rasa/notify/commits/main/) and ran the tests [again](https://github.com/rasa/notify/actions) but the code fails the tests almost all the time, except on Windows (when not testing against the root folder).

Unfortunately, I haven't made much progress other than building a better test harness. Note that this issue was [identified](https://github.com/rjeczalik/notify/issues/148) over six years ago, several people worked on fixing it (other than me) to no avail, and upstream hasn't been updated in over two years.

I would suggest we wait and see if https://github.com/fsnotify/fsnotify/pull/676 gets merged in. If so, it may be worth seeing if we can switch to it, since it's actively maintained, and [appears](https://github.com/fsnotify/fsnotify?tab=readme-ov-file) to support all the platforms we do.

**Comment by rasa** [comment]
@MU7L3Y Using a junction, or a mount point, you can work around this issue. So, let's say you wanna sync `D:\`.

Run
```
mklink /J C:\D D:\
```
Then sync `C:\D` and it should work.

**Comment by tomasz1986** [comment]
@rasa Are you 100% sure about this? We've done tests on the forum in https://forum.syncthing.net/t/no-istant-sync/17497 (albeit a few years ago), and the watcher did not work even when trying to work around the problem using junctions or mount points.

**Comment by rasa** [comment]
> [@rasa](https://github.com/rasa) Are you 100% sure about this? 

@tomasz1986 It was months ago, so no, not 100% sure. I will test it again. I only tried the Disk Management app, but I assume if that works a junction would work as well. I will test that too.

I know the opposite causes the code to fail, so [if we](https://github.com/syncthing/notify/commit/581b9618a899c4f92eb28f1155fec508a7d2f109#diff-245392b692a50c38ecab4381b118862db514035c10983f3bd4f4b7f1f4be4692R108) `SUBST Z: C:\tmp` and then watch `Z:\`, the tests [fail](https://github.com/rasa/notify/actions/runs/13680925950).

**Comment by MU7L3Y** [comment]
@rasa I actually tried `mklink /D` the other day but it seems it sometimes works and sometimes doesn't. I don't think it's reliable. I've had to drop my full rescan of over one million files to 20mins to ensure users get changes in not too long a time although not ideal.

**Comment by rasa** [comment]
Or if CGO is now an option, https://github.com/dunglas/go-fswatch might be a better choice.
